### PR TITLE
SPLAT-636: enabling AWS Local Zones in BYO VPC (phase-1)

### PR DESCRIPTION
**Pull request by mtulio**
AWS Local Zones Phase-1: Install a cluster in existing VPC with Local Zone subnets (on install-config.yaml).

Some fragments will be adapted to existing steps if it's applied.

https://issues.redhat.com/browse/SPLAT-728
https://issues.redhat.com/browse/SPLAT-636 (feature)

Depends on:

- https://github.com/openshift/release/pull/37543
- https://github.com/openshift/ci-tools/pull/3336
- https://github.com/openshift/ci-tools/pull/3339
- https://github.com/openshift/release/pull/37718

### CODE REVIEW

**Comment by mtulio** [comment]
cc @rvanderp3 
/hold

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:1]: should be adapted with `ci-operator/step-registry/aws/provision/localzone-subnets/aws-provision-localzone-subnets-commands.sh`

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-ref.yaml
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-ref.yaml:5]:
```suggestion
    name: "4.13"
```

File: ci-operator/step-registry/openshift/e2e/aws/localzone/byo-vpc/openshift-e2e-aws-ovn-workflow.yaml
Comment (mtulio) [inline ci-operator/step-registry/openshift/e2e/aws/localzone/byo-vpc/openshift-e2e-aws-ovn-workflow.yaml:7]: we may need to add a custom LZ tests here, for example:
- check if LZ node is running with expected labels
- check if MTU is set correctly
- create an app on LZ, using the node selectors for it (make sure the image uses the internal registry)
- quickly measure the latency between zones in the region and LZ by curl-ing from nodes in the region

File: ci-operator/step-registry/openshift/e2e/aws/localzone/openshift-e2e-aws-ovn-workflow.yaml
Comment (mtulio) [inline ci-operator/step-registry/openshift/e2e/aws/localzone/openshift-e2e-aws-ovn-workflow.yaml:1]: I am not sure if it's correct.

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:151]: we need to make sure the credentials will have permissions

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:170]:
```suggestion
echo $localzone_subnet > $SHARED_DIR/localzone_subnet_id
echo $localzone_subnet >> "${SHARED_DIR}/subnet_ids"
```

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:130]:
```suggestion
if [[ ${PUBLIC_SUBNET} == true ]]; then
```

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-ref.yaml
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-ref.yaml:14]:
```suggestion
  - name: PUBLIC_SUBNET
```

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:111]: Why not getting from the installer repo. Some examples:

https://github.com/openshift/release/blob/eee207c1fb11c489da92ac88ab6fa87a9b04ce38/ci-operator/step-registry/ipi/conf/aws/sharednetwork/ipi-conf-aws-sharednetwork-commands.sh

**Comment by mtulio** [comment]
> time="2023-03-21T23:14:36Z" level=error msg="Failed to load CI Operator configuration" error="invalid ci-operator config: invalid configuration: tests[73]: invalid cluster profile \"aws-local-zones\"" source-file=/ci-operator/config/openshift/release/openshift-release-master__nightly-4.13.yaml

Waiting https://github.com/openshift/ci-tools/pull/3336 to be merged

**Comment by mtulio** [comment]
https://github.com/openshift/release/pull/37543 and https://github.com/openshift/ci-tools/pull/3336 merged. Converting to regular PR

**Comment by mtulio** [comment]
Richard, would you mind to take a look?
/assign @rvanderp3

**Comment by mtulio** [comment]
/pj-rehearse  periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

File: ci-operator/jobs/openshift/release/openshift-release-master-periodics.yaml
Comment (mtulio) [inline ci-operator/jobs/openshift/release/openshift-release-master-periodics.yaml:47322]:
```suggestion
```

**Comment by mtulio** [comment]
Job recreated after the ci-tools change (disabling ConfigMap https://github.com/openshift/ci-tools/pull/3339 ).
Awaiting for CI pass to rehearsal again:

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
The identified issue on jq parser when extracting the public route table ID from the VPC-Shared step (CloudFormation stack). Fixed and trying again:

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
SSH public key was added to the secret mounted to cluster-profile:

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
```
cp: cannot stat '/var/run/secrets/ci.openshift.io/cluster-profile/ssh-privatekey': No such file or directory
```

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
- Local Zone subnet merged to the subnet IDs

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
> /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

Cfn templates created on the zones a, b, and c, the scripts wrap the compute pool to two zones randomly chosen, when choosing the zone not created by the cloudformation template, it is expected to fail. It should be addressed in this job, and I think overall on CI.

```
level=error msg=failed to fetch Master Machines: failed to load asset "Install Config": failed to create install config: [controlPlane.platform.aws.zones: Invalid value: []string{"us-west-2b", "us-west-2d"}: No subnets provided for zones [us-west-2d], compute[0].platform.aws.zones: Invalid value: []string{"us-west-2b", "us-west-2d"}: No subnets provided for zones [us-west-2d]]
```

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
> * Local Zone subnet merged to the subnet IDs
> 
> /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

it seems there are failures on e2e related to the CSI on Local Zones. Need to run more jobs to understand the pattern or if it's an persistent issue.

```
[sig-storage] In-tree Volumes [Driver: aws] [Testpattern: Pre-provisioned PV (default fs)] volumes should allow exec of files on the volume [Skipped:NoOptionalCapabilities] [Suite:openshift/conformance/parallel] [Suite:k8s] expand_less | 5m5s
-- | --
{  fail [k8s.io/kubernetes@v1.26.1/test/e2e/storage/drivers/in_tree.go:1604]: Mar 28 20:57:08.308: RequestError: send request failed caused by: Post "https://ec2.us-east-1-dfw-1.amazonaws.com/": dial tcp: lookup ec2.us-east-1-dfw-1.amazonaws.com on 172.30.0.10:53: no such host Ginkgo exit error 1: exit with code 1}
```

Ref  https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/36911/rehearse-36911-periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc/1640806152029605888

**Comment by mtulio** [comment]
> /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

This must be fixed:
```
No subnets provided for zones [us-east-1f], compute[0].platform.aws.zones: Invalid value: []string{"us-east-1c", "us-east-1f"}: No subnets provided for zones [us-east-1f]]
```

https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/36911/rehearse-36911-periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc/1640931157438631936#1:build-log.txt%3A49

**Comment by mtulio** [comment]
The zone names now is discovered on the Local Zone step and set to the IC, preventing the further steps (`ipi-conf-aws`) to force zones on IC that is not created.

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
> /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

Install failed, points to cloud infra flaky

/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
> > /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc
> 
> Install failed, points to cloud infra flaky
> 
> /pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

Successful execution after disabling the e2e tests (well, not the goal of this PR, but I needed to validate the flow).

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

**Comment by mtulio** [comment]
Only network tests have been skipped :
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-localzone-byo-vpc

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh:125]:
```suggestion
# Choosing randomly the AZ within the Region (best choice to test any AZ - and detect possible errors)
```

**Comment by mtulio** [comment]
tests passed!

Job renamed to fit the feature name (Local Zone**s**)

Bug also opened to address the e2e skips. https://issues.redhat.com/browse/OCPBUGS-11609

/pj-rehearse refresh

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-local-zones-byo-vpc

**Comment by mtulio** [comment]
/pj-rehearse periodic-ci-openshift-release-master-nightly-4.13-e2e-aws-ovn-local-zones-byo-vpc

File: ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml
Comment (jianlinliu) [inline ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: Seem like localzone development was already done in https://github.com/openshift/release/pull/37022, can we use the same provision/pre chain on different cluster profile to run the ci? cc @yunjiang29

**Review by vrutkovs — APPROVED** [review]
/approve

File: ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml
Comment (mtulio) [reply ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: Yes, this LZ deployment was based on #37022 but will be changed in the next phase (II).
there are two main differences from the goals for this deployment:
1) on phase II (4.14) we will use the CloudFormation template provided on the installer repo, instead of embedding it [it was not added on this PR as we need to run this periodic asap to capture signals of phase I/4.13 deployment]
2) the current LZ deployment is selecting randomly the zone by default, increasing the locations to test. In my last commit, I ported the existing deployment to keep more compatibility [but not yet tested]. The goal is to talk with @yunjiang29 when 1 is finished: https://github.com/openshift/release/compare/e97a4db64b6afac2aaa015ee270ce46def9b5613..e0dbe15a1cce3dee1858ba044b2df4c1fee2746c

**Comment by mtulio** [comment]
> Hold for a while, to discuss if we can share the same install chains between QE and Dev.

/assign @yunjiang29

File: ci-operator/step-registry/openshift/e2e/aws/localzone/byo-vpc/openshift-e2e-aws-ovn-workflow.yaml
Comment (mtulio) [reply ci-operator/step-registry/openshift/e2e/aws/localzone/byo-vpc/openshift-e2e-aws-ovn-workflow.yaml]: more coverage will be implemented on https://issues.redhat.com/browse/SPLAT-1028

File: ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh
Comment (mtulio) [reply ci-operator/step-registry/aws/provision/localzone/aws-provision-localzone-commands.sh]: https://issues.redhat.com/browse/SPLAT-1028

File: ci-operator/step-registry/ipi/conf/aws/local-zones/byo-vpc/ipi-conf-aws-local-zones-byo-vpc-ref.yaml
Comment (yunjiang29) [inline ci-operator/step-registry/ipi/conf/aws/local-zones/byo-vpc/ipi-conf-aws-local-zones-byo-vpc-ref.yaml:2]: @mtulio Tasks in this step are currently done by `ipi-conf-aws-localzone` + `ipi-conf-aws-custom-vpc`,  I still think it's better to keep existing logic:
`ipi-conf-aws-custom-vpc` -> set subnet + zones
`ipi-conf-aws-localzone` -> set `compute[?name=edge]` + lz subnet,  and we can add more params in this step later

File: ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml
Comment (mtulio) [inline ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:17]:
```suggestion
```

File: ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/aws/provision/vpc/shared/aws-provision-vpc-shared-commands.sh:387]:
```suggestion
# ['private-rtb1', 'private-rtb2']
```

File: ci-operator/step-registry/ipi/install/install/aws/ipi-install-install-aws-commands.sh
Comment (mtulio) [inline ci-operator/step-registry/ipi/install/install/aws/ipi-install-install-aws-commands.sh:154]:
```suggestion
```

**Review by mtulio — COMMENTED** [review]
some left comments. Will wait for the rehearsals to remove it.

File: ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml
Comment (yunjiang29) [reply ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: @mtulio thanks for adding extra test points, I noticed some codes are duplicated as existing code, that would be better if we could re-use them instead of creating new steps.

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: Thanks for your follow up. 

Some background that QE migrate testing from internal jenkins to prow, one important goal is QE and Dev can share the same steps in testing, and do contributions for a unified ci, enhance the collaboration between QE and Dev to work more closer together, also can help keep us on the same page. This is our KR goals. Re-using the same would also help eliminate some cases that someone can reproduce issues, but others can not.

That is why we proposed to re-use the new steps as much as possible.

Comment (mtulio) [reply ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: > Thanks for your follow up.
> 
> Some background that QE migrate testing from internal jenkins to prow, one important goal is QE and Dev can share the same steps in testing, and do contributions for a unified ci, enhance the collaboration between QE and Dev to work more closer together, also can help keep us on the same page. This is our KR goals. Re-using the same would also help eliminate some cases that someone can reproduce issues, but others can not.
> 
> That is why we proposed to re-use the new steps as much as possible.

💯 Yeah, that makes total sense to me and I will work to improve that. But I am not sure how to troubleshoot those [failing jobs without access to it](https://github.com/openshift/release/pull/36911#issuecomment-1507927121) - mainly when refactoring/adapting the steps maintained by QE to the use case we are implementing.

I am open to proposals on how to reuse tools that can unblock both teams.

cc @rvanderp3

Comment (jianlinliu) [reply ci-operator/step-registry/ipi/aws/pre/local-zones/byo-vpc/ipi-aws-pre-local-zones-byo-vpc-chain.yaml:2]: > I am not sure how to troubleshoot those https://github.com/openshift/release/pull/36911#issuecomment-1507927121
@mtulio after confirmed with QE team, you can refer to FAQ-2 in https://docs.google.com/document/d/1qOxZ33lWfZRFaJC5b0U4sz3VMseMsXfTJ-1ilXgdFm4/edit to get the permission.

**Comment by mtulio** [comment]
closing this card in favor of reusing existing CloudFormation Templates and steps when provisioning existing shared-vpc workflow https://github.com/openshift/release/pull/39902

**Pull request closed without merging**
